dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,182 @@
+)
+
+
+urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+
+class KeyCloakError(DLabException):
+    def __init__(self, *args, **kwargs):
+        self.message = self.message.format(*args, **kwargs)
+
+    def __str__(self):
+        return self.message
+
+
+class InvalidParameterFormatError(KeyCloakError):
+    message = 'Parameter {value} is not of {expected_type} type: {actual_type}'
+
+
+class ConnectionError(KeyCloakError):
+    message = 'Server responded with {status_code} - {content}'
+
+
+def check_response(func):
+    """Wrapper that checks if response returned by wrapped function is good
+
+    :type func: function
+    :param func: function to wrap
+
+    :rtype: requests.Response
+    :return: :class:`Response <Response>` object
+
+    :raises InvalidResponseError: if response status code is not 200
+    """
+    def wrapper(*args, **kwargs):
+        response = func(*args, **kwargs)
+        if not response.ok:
+            raise ConnectionError(
+                status_code=response.status_code,
+                content=response.content.decode('utf8')
+            )
+        return response
+
+    return wrapper
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use (e.g.
+`https://12.12.12.12/auth/`)
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        self.check_args_types(locals(), str)
+        self._realm_address = '{}realms/{}'.format(keycloak_host, realm_name)
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self.logger = StreamLogBuilder('keycloak', INFO).logging
+        self._pub_key = self.get_key()
+
+    def check_args_types(self, args, expected_type):
+        del args['self']
+        for arg_name, arg_value in args.items():
+            if not isinstance(arg_value, expected_type):
+                raise InvalidParameterFormatError(
+                    value=arg_name,
+                    expected_type=expected_type.__name__,
+                    actual_type=type(arg_value).__name__
+                )
+
+    @check_response
+    def make_request_to_server(self, method='GET', endpoint='', **kwargs):
+        """Performs request to KeyCloak server and checks response
+
+        :type method: str
+        :param method: method for the HTTP request
+        :type endpoint: str
+        :param endpoint: KeyCloak server endpoint to make request to
+
+        :rtype: requests.Response
+        :return: HTTP response object
+        """
+        kwargs['verify'] = True
+        response = requests.request(method, endpoint, **kwargs)
```

Comment:
you need to handle request execution flow here

Reply:
Added a check for any request error after receiving response